src/lib/rng/chacha_rng/chacha_rng.cpp

```diff
@@ -58,25 +58,20 @@ void ChaCha_RNG::clear_state()
    m_chacha->set_key(m_hmac->final());
    }
 
-void ChaCha_RNG::generate_output(uint8_t output[], size_t output_len,
-                                 const uint8_t input[], size_t input_len)
+void ChaCha_RNG::generate_output(std::span<uint8_t> output, std::span<const uint8_t> input)
    {
-   if(input_len > 0)
+   if(!input.empty())
       {
-      update(input, input_len);
-      }
-
-   m_chacha->write_keystream(output, output_len);
-   }
+      m_hmac->update(input.data(), input.size()); // TODO: fix after GH #3294
+      m_chacha->set_key(m_hmac->final());
 
-void ChaCha_RNG::update(const uint8_t input[], size_t input_len)
-   {
-   m_hmac->update(input, input_len);
-   m_chacha->set_key(m_hmac->final());
+      secure_vector<uint8_t> mac_key(m_hmac->output_length());
+      m_chacha->write_keystream(mac_key.data(), mac_key.size());
+      m_hmac->set_key(mac_key);
+      }
 
-   secure_vector<uint8_t> mac_key(m_hmac->output_length());
-   m_chacha->write_keystream(mac_key.data(), mac_key.size());
-   m_hmac->set_key(mac_key);
```

Comment:
Previously, `Stateful_RNG` had `::update()` and `::generate_output()`. IIUC that was done so that `add_entropy()` could just call `::update()` and don't bother with the generation code at all. Now that `add_entropy` is implemented in terms of `::fill_bytes_with_input()` we don't need that distinction anymore. Merging `::update()` and `::generate_output()` increases the control flow complexity a little, though.